### PR TITLE
Ggplot

### PR DESCRIPTION
**Pull request by MartinHinz**
Added ggplot output functionality to the plots. This is the last open issue for milestone oxcAAR 1.1 Flora, so it might be worthwhile checking it in more detail

### CODE REVIEW

File: R/oxcAARCalibratedDate.R [hunk 1]
```diff
@@ -81,13 +81,138 @@ format.oxcAARCalibratedDate <- function(x, ...){
...
+
+  this_sigma_ranges <- x$sigma_ranges
+  this_sigma_qualifier <- "unmodelled"
+  if(all(!(is.null(x$posterior_sigma_ranges)), !(is.na(x$posterior_sigma_ranges)))) {
+    this_sigma_ranges <- x$posterior_sigma_ranges
+    this_sigma_qualifier <- "modelled"
+  }
+
+  sigma_text<-paste(this_sigma_qualifier,
+                    oxcAAR:::formatFullSigmaRange(this_sigma_ranges$one_sigma,"one sigma"),
```
Comment (nevrome) [inline R/oxcAARCalibratedDate.R:184]: Just as the CRAN checks say:
```
  There are ::: calls to the package's namespace in its code. A package
    almost never needs to use ::: for its own objects
```
Usually you can call unexported package functions without a namespace definition, so only `formatFullSigmaRange()` is fine. Or was this for development? I recommend `devtools::load_all()` in that case.

File: DESCRIPTION [hunk 2]
```diff
@@ -15,9 +15,12 @@ LazyData: TRUE
 Imports:
   stringi,
   stringr,
-  jsonlite
+  jsonlite,
+  ggplot2,
```
Comment (nevrome) [inline DESCRIPTION:19]: Heavy dependencies, especially when there are very few packages necessary so far. Maybe it's better to make this whole ggplot thing optional. I think we had this solution for mortAAR already, where ggplot2 is only in the suggests.

File: NAMESPACE [hunk 3]
```diff
@@ -53,7 +54,26 @@ export(quickSetupOxcal)
 export(readOxcalOutput)
 export(setOxcalExecutablePath)
 export(wrap_in_boundaries)
+importFrom("ggplot2","aes")
```
Comment (nevrome) [inline NAMESPACE:57]: Clever to import these here. But again: Maybe better to avoid this heavy dependency.

File: NAMESPACE [hunk 4]
```diff
@@ -31,6 +31,7 @@ export(Phase)
 export(R_Date)
 export(R_Simulate)
 export(Sequence)
+export(as.CalDates)
```
Comment (nevrome) [inline NAMESPACE:34]: Sneaky! That should be in another PR :smile:

File: R/oxcAARCalibratedDate.R [hunk 5]
```diff
@@ -81,13 +81,138 @@ format.oxcAARCalibratedDate <- function(x, ...){
...
+  } else {
+    plotoxcAARDateSystemGraphics(x, ...)
+  }
 }
+
+#' @importFrom "ggplot2" "ggplot" "aes" "geom_area" "labs" "theme_light" "geom_ribbon" "scale_y_continuous" "sec_axis" "ggplot_build" "geom_polygon" "scale_x_continuous" "annotate" "geom_errorbarh" "scale_alpha_manual"
+#' @importFrom "stats" "dnorm"
+plotoxcAARDateGGPlot2<-function(x, ...){
+  bc <- .data <- NULL
+  to_plot <- data.frame(dates=x$raw_probabilities$dates,
```
Comment (nevrome) [inline R/oxcAARCalibratedDate.R:96]: Horrible code layout, but you do you. 

I don't see anything wrong with the implemented logic :+1:

File: R/oxcAARCalibratedDate.R [hunk 6]
```diff
@@ -206,6 +331,71 @@ plotoxcAARDateSystemGraphics <- function(x, ...){
...
+    sublist <- vector(mode="list", length=length(x))
+    names(sublist) <- names(x)
+    names(reslist) <- c("metadata","grids")
+    ## metadata
+    df <- as.data.frame(matrix(ncol=12, nrow=length(x)), stringsAFactors=FALSE)
+    names(df) <- c("DateID","CRA","Error","Details","CalCurve","ResOffsets","ResErrors","StartBP","EndBP","Normalised","CalEPS", "Method")
+    df$DateID <- names(x)
+    df$CRA <- as.numeric(unlist(lapply(X=x, FUN=`[[`, "bp")))
+    df$Error <- as.numeric(unlist(lapply(X=x, FUN=`[[`, "std")))
+    df$CalCurve=lapply(lapply(lapply(lapply(lapply(x,FUN=`[[`,"cal_curve"),FUN=`[[`,"name"),strsplit," "),unlist),FUN=`[[`,2)
```
Comment (nevrome) [inline R/oxcAARCalibratedDate.R:371]: `lapply(lapply(lapply(lapply(lapply` - really? There must be a better readable way to express this!

File: R/oxcAARCalibratedDate.R [hunk 7]
```diff
@@ -206,6 +331,71 @@ plotoxcAARDateSystemGraphics <- function(x, ...){
...
+      res$PrDens <- res$PrDens/sum(res$PrDens)
+      class(res) <- append(class(res),"calGrid")
+      sublist[[i]] <- res
+    }
+    reslist[["grids"]] <- sublist
+    reslist[["calmatrix"]] <- NA
+    class(reslist) <- c("CalDates",class(reslist))
+    return(reslist)
+  }
+}
```
Comment (nevrome) [inline R/oxcAARCalibratedDate.R:397]: Not sure if this is perfectly clean.

```
> my_uncal_dates <- data.frame(bp=c(5000,4500,4000),
                             std=c(45,35,25),
                             names=c("Date 1", "Date 2", "Date 3")
                             )
> my_cal_dates <- oxcalCalibrate(my_uncal_dates$bp, my_uncal_dates$std, my_uncal_dates$names)
> as.CalDates(my_cal_dates) %>% str()
List of 3
 $ metadata :'data.frame':	3 obs. of  11 variables:
  ..$ DateID    : chr [1:3] "Date 1" "Date 2" "Date 3"
  ..$ CRA       : num [1:3] 5000 4500 4000
  ..$ Error     : num [1:3] 45 35 25
  ..$ Details   : logi [1:3] NA NA NA
  ..$ CalCurve  : chr [1:3] "data" "data" "data"
  ..$ ResOffsets: logi [1:3] NA NA NA
  ..$ ResErrors : logi [1:3] NA NA NA
  ..$ StartBP   : logi [1:3] NA NA NA
  ..$ EndBP     : logi [1:3] NA NA NA
  ..$ Normalised: logi [1:3] TRUE TRUE TRUE
  ..$ CalEPS    : num [1:3] 0 0 0
 $ grids    :List of 3
  ..$ Date 1:Classes ‘calGrid’ and 'data.frame':	545 obs. of  2 variables:
  .. ..$ calBP : num [1:545] 6009 6008 6007 6006 6005 ...
  .. ..$ PrDens: num [1:545] 0 0 0 0 0 0 0 0 0 0 ...
  ..$ Date 2:Classes ‘calGrid’ and 'data.frame':	635 obs. of  2 variables:
  .. ..$ calBP : num [1:635] 5474 5473 5472 5471 5470 ...
  .. ..$ PrDens: num [1:635] 0 0 0 0 0 ...
  ..$ Date 3:Classes ‘calGrid’ and 'data.frame':	585 obs. of  2 variables:
  .. ..$ calBP : num [1:585] 4814 4813 4812 4811 4810 ...
  .. ..$ PrDens: num [1:585] 0 0 0 0 0 ...
 $ calmatrix: logi NA
 - attr(*, "class")= chr [1:2] "CalDates" "list"
```
```
> rcarbon::calibrate(c(5000,4500,4000), errors=c(45,35,25)) %>% str()
[1] "Calibrating radiocarbon ages..."
  |=====================================================================================================| 100%
[1] "Done."
List of 3
 $ metadata :'data.frame':	3 obs. of  12 variables:
  ..$ DateID    : chr [1:3] "1" "2" "3"
  ..$ CRA       : num [1:3] 5000 4500 4000
  ..$ Error     : num [1:3] 45 35 25
  ..$ Details   : logi [1:3] NA NA NA
  ..$ CalCurve  : chr [1:3] "intcal20" "intcal20" "intcal20"
  ..$ ResOffsets: num [1:3] 0 0 0
  ..$ ResErrors : num [1:3] 0 0 0
  ..$ StartBP   : num [1:3] 55000 55000 55000
  ..$ EndBP     : num [1:3] 0 0 0
  ..$ Normalised: logi [1:3] TRUE TRUE TRUE
  ..$ F14C      : logi [1:3] FALSE FALSE FALSE
  ..$ CalEPS    : num [1:3] 1e-05 1e-05 1e-05
 $ grids    :List of 3
  ..$ 1:Classes ‘calGrid’ and 'data.frame':	336 obs. of  2 variables:
  .. ..$ calBP : num [1:336] 5922 5921 5920 5919 5918 ...
  .. ..$ PrDens: num [1:336] 1.11e-05 1.29e-05 1.49e-05 1.77e-05 2.10e-05 ...
  ..$ 2:Classes ‘calGrid’ and 'data.frame':	443 obs. of  2 variables:
  .. ..$ calBP : num [1:443] 5432 5431 5430 5429 5428 ...
  .. ..$ PrDens: num [1:443] 1.10e-05 1.27e-05 1.47e-05 1.42e-05 1.37e-05 ...
  ..$ 3:Classes ‘calGrid’ and 'data.frame':	270 obs. of  2 variables:
  .. ..$ calBP : num [1:270] 4611 4610 4609 4608 4607 ...
  .. ..$ PrDens: num [1:270] 1.19e-05 1.35e-05 1.91e-05 2.14e-05 2.14e-05 ...
 $ calmatrix: logi NA
 - attr(*, "class")= chr [1:2] "CalDates" "list"
```
There is some factual loss of information, e.g. in the `CalCurve` field.

File: R/oxcAARCalibratedDatesList.R [hunk 8]
```diff
@@ -19,18 +19,64 @@ print.oxcAARCalibratedDatesList <- function(x, ...){
...
+  graph <- m + geom_ridgeline(aes(x = .data$dates,
+                                  y = .data$name,
+                                  height = .data$probability,
+                                  alpha = .data$alpha,
+                                  group = interaction(.data$class,.data$name)),
+                              scale=100,
+                              fill = "#fc8d62",
+                              color="#00000077") +
+    theme_light() + labs(y="Dates")  +
+    ggplot2::scale_alpha_continuous(guide = FALSE)
```
Comment (nevrome) [inline R/oxcAARCalibratedDatesList.R:76]: Why no `importFrom` here?

File: R/oxcAARCalibratedDate.R [hunk 9]
```diff
@@ -81,13 +81,138 @@ format.oxcAARCalibratedDate <- function(x, ...){
...
+  this_sigma_qualifier <- "unmodelled"
+  if(all(!(is.null(x$posterior_sigma_ranges)), !(is.na(x$posterior_sigma_ranges)))) {
+    this_sigma_ranges <- x$posterior_sigma_ranges
+    this_sigma_qualifier <- "modelled"
+  }
+
+  sigma_text<-paste(this_sigma_qualifier,
+                    oxcAAR:::formatFullSigmaRange(this_sigma_ranges$one_sigma,"one sigma"),
+                    oxcAAR:::formatFullSigmaRange(this_sigma_ranges$two_sigma,"two sigma"),
+                    oxcAAR:::formatFullSigmaRange(this_sigma_ranges$three_sigma,"three sigma"),
```
Comment (nevrome) [inline R/oxcAARCalibratedDate.R:186]: Do you really need 6 decimal places for the %-value? Maybe 2 would be enough?

File: R/oxcAARCalibratedDate.R [hunk 10]
```diff
@@ -81,13 +81,138 @@ format.oxcAARCalibratedDate <- function(x, ...){
...
+              fill = "#fc8d62",
+              position = "identity",
+              color="#00000077"
+              ) +
+    labs(title = paste0(x$name,
+                        ": ",
+                        x$bp,
+                        "\u00B1",
+                        x$std),
+         caption = x$cal_curve$name,
```
Comment (nevrome) [inline R/oxcAARCalibratedDate.R:149]: Maybe there is a way to add the more obvious "IntCal20" somewhere to make more clear, what "Atmospheric data from Reimer et al (2020)" means? I guess that's just what oxcal returns?

**Review by nevrome — APPROVED** [review]
Good job - as always! But I find your code layout difficult to read. A more vertical style might be better.

**Comment by MartinHinz** [comment]
Thank you for your very swift and detailed review! I see that there is still quite some room for improvement and I will correct the pull request accordingly before merging. Might be that I will ask you again for review this and the final release, if you are willing to do so!

File: R/oxcAARCalibratedDate.R (hunk 1 above)
Comment (MartinHinz) [reply R/oxcAARCalibratedDate.R:184]: I thought, I had a good reason for this... But I removed it, works also like this, so consider it as solved.

File: DESCRIPTION (hunk 2 above)
Comment (MartinHinz) [reply DESCRIPTION:19]: I absolutely agree. But putting the packages under suggests results in an 

```
> checking package dependencies ... ERROR
  Namespace dependencies not required: 'ggplot2', 'ggridges', 'methods'
```

Any suggestions appreciated!

File: NAMESPACE (hunk 3 above)
Comment (MartinHinz) [reply NAMESPACE:57]: See above, results from the doxygen like @imports. Do you think it is save to remove them?

File: NAMESPACE (hunk 4 above)
Comment (MartinHinz) [reply NAMESPACE:34]: Right, is part of another branch and sneaked its way into this one. Removed from this branch

File: NAMESPACE (hunk 3 above)
Comment (nevrome) [reply NAMESPACE:56]: If you want to remove them to keep ggplot2 etc. in the Suggests, you have to call each of these functions explicitly in the code, so with their respective namespace (`ggplot2::aes`, etc.).

You should also check if the packages you only have in the Suggests are available at the beginning of the function with something like:

```r
check_if_packages_are_available <- function(packages_ch) {
  if (
    packages_ch %>%
    sapply(function(x) {requireNamespace(x, quietly = TRUE)}) %>%
    all %>% `!`
  ) {
    stop(
      paste0(
        "R packages ",
        paste(packages_ch, collapse = ", "),
        " needed for this function to work. Please install with ",
        "install.packages(c('", paste(packages_ch, collapse = "', '"), "'))"
      ),
      call. = FALSE
    )
  }
}
```

Otherwise the user will get pretty weird feedback.

File: DESCRIPTION (hunk 2 above)
Comment (nevrome) [reply DESCRIPTION:19]: See below.

File: R/oxcAARCalibratedDate.R (hunk 6 above)
Comment (MartinHinz) [reply R/oxcAARCalibratedDate.R]: Not part of this pr any longer, has to be discussed again at the appropriate place ;-)

File: R/oxcAARCalibratedDate.R (hunk 7 above)
Comment (MartinHinz) [reply R/oxcAARCalibratedDate.R]: made an issue for the rcarbon-sync-branch

File: R/oxcAARCalibratedDatesList.R (hunk 8 above)
Comment (MartinHinz) [reply R/oxcAARCalibratedDatesList.R:75]: Forgotten, now done.

File: R/oxcAARCalibratedDate.R (hunk 9 above)
Comment (MartinHinz) [reply R/oxcAARCalibratedDate.R]: Thank you! That is a general issue not directly related to this PR, I opened a special issue for that.

File: R/oxcAARCalibratedDate.R (hunk 10 above)
Comment (MartinHinz) [reply R/oxcAARCalibratedDate.R:149]: Good point. Nevertheless, this takes the result directly from the oxcal response, I would rather keep this behavior?

**Comment by MartinHinz** [comment]
updated the PR. Would you, @nevrome, be probably so kind as rechecking, resolving and advising me on the still open points. Much appreciated!

**Comment by MartinHinz** [comment]
Also, forgot to merge master back in. Doing this in this minute and push when checked. Sorry!

File: NAMESPACE (hunk 3 above)
Comment (nevrome) [reply NAMESPACE:64]: I almost never use the `importFrom` notation, so I don't know how it behaves. But I assume it's the reason why you can't put ggplot2 into the Suggests without triggering
```
Namespace dependencies not required: 'ggplot2', 'ggridges', 'methods'
```
Probably you could go with the `::` notation to avoid this.

I'm not 100% sure if I'm missing something here. You could check our solutions in c14bazAAR and mortAAR for a working example with a lot of the supplementary packages in the Suggests field.

Comment (MartinHinz) [reply NAMESPACE]: With e7e402f now this should be solved?

File: DESCRIPTION (hunk 2 above)
Comment (MartinHinz) [reply DESCRIPTION]: see below.

File: R/oxcAARCalibratedDate.R (hunk 10 above)
Comment (nevrome) [reply R/oxcAARCalibratedDate.R:147]: OK :shrug:

**Comment by nevrome** [comment]
I think this is good to go :+1:

**Comment by MartinHinz** [comment]
Thank you so much! Merging now.
